devdoc/channel_2.md

```diff
@@ -0,0 +1,72 @@
+# channel_2 requirements:
+
+## Overview
+
+The current implementation of `channel` has certain pitfalls. `channel_2` aims to address these pitfalls by making the following changes:
```

Comment:
>certain pitfalls [](http://example.com/codeflow?start=44&length=16)

there is another pitfall that isn't addressed here. With the current code, there is now way for the caller to know when the channel is actually closed.  The callbacks all happen in threadpool workers, but after you cancel an operation, there is no way to know when those workers are done. I talk about this in earlier comments, but I think it's really important to be very crisp about when functions like "cancel" and "close" return. If something queues a callback and immediately returns, then that is not crisp.